src/ui-scroll.component.ts

```diff
@@ -46,11 +80,14 @@ export class UiScrollComponent implements OnInit, OnDestroy {
 
   constructor(
     public changeDetector: ChangeDetectorRef,
-    public elementRef: ElementRef) { }
+    public elementRef: ElementRef
+  ) {
+    setTimeout(() => this.ngOnInit()); // 😢
+  }
```

Comment:
@markdBC  To say I don't like this would be a considerable understatement... ngOnInit must be invoked implicitly, and I have no idea why it became broken. It has something to do with new approach of creating the component on the directive level. The [demo we developed on stackblitz](https://stackblitz.com/edit/angular-structural-dir-with-comp-table) has the same issue.

It would be very nice if you could take a look and help with it. For testing in ngx-ui-scroll environment you may use http://localhost:4200/#/test (after npm start), which runs demo/app/samples/test.component.html template. There are multiple issues behind this one, so to see the result is still impossible, but the initialization could be handeled in a proper way

Reply:
The problem is that `ngOnInit` is called after the first time `ngOnChanges` is called (https://angular.io/guide/lifecycle-hooks#lifecycle-sequence).

However, while `ViewContainerRef.CreateComponent` causes an invocation of `ngOnChanges`, the same is not true for `ComponentRef.Create`.

You can find the definition of `ViewContainerRef.CreateComponent`, at `packages/core/src/view/refs.ts:180`  of the Angular source code:

```
  createComponent<C>(
      componentFactory: ComponentFactory<C>, index?: number, injector?: Injector,
      projectableNodes?: any[][], ngModuleRef?: NgModuleRef<any>): ComponentRef<C> {
    const contextInjector = injector || this.parentInjector;
    if (!ngModuleRef && !(componentFactory instanceof ComponentFactoryBoundToModule)) {
      ngModuleRef = contextInjector.get(NgModuleRef);
    }
    const componentRef =
        componentFactory.create(contextInjector, projectableNodes, undefined, ngModuleRef);
    this.insert(componentRef.hostView, index);
    return componentRef;
  }
```

`ViewContainerRef.CreateComponent` calls the `ComponentRef.Create` method, and then calls `ViewContainerRef.Insert`, which among other things causes change detection to occur inside the component.

A solution is to explicitly invoke change detection for the component, via `ComponentRef.changeDetectorRef.detectChanges()`. This is probably only slightly better than explicitly invoking `ngOnInit`, but at least we're not calling a lifecycle hook directly.

I've made a PR with this change at https://github.com/dhilt/ngx-ui-scroll/pull/147 and have a StackBlitz demo at https://stackblitz.com/edit/angular-structural-dir-with-comp-table-rqlfvk?file=src%2Fapp%2FmyDir.ts.